sdk/metric/internal/sum.go

```diff
@@ -260,3 +267,8 @@ type precomputedSum[N int64 | float64] struct {
 func (s *precomputedSum[N]) Aggregate(value N, attr attribute.Set) {
 	s.set(value, attr)
 }
+
+// aggregateFiltered records value with spatially re-aggregated attrs.
+func (s *precomputedSum[N]) aggregateFiltered(value N, attr attribute.Set) { // nolint: unused  // used to filter.
+	s.valueMap.Aggregate(value, attr)
```

Comment:
I thought precomputed sums were supposed to be reset each time they are collected?  Will this never be reset?

Reply:
For cumulative precomputed sums, it is the same as cumulative sums, they do not reset as the cumulative value is considered still relevant the next collection cycle. For delta cumulative sums, they do.

This change does not modify that behavior though. The cumulative sums preserve the previous value and the delta reset when `Aggregation` are called. The line you commented on is a shorthand for invoking this: https://github.com/open-telemetry/opentelemetry-go/blob/a724cf884287e04785eaa91513d26a6ef9699288/sdk/metric/internal/sum.go#L41-L45 which does not affect the collection reset functionality.

Reply:
This is effectively replacing the line above, since we would previously call Aggregate from sdk/metric/internal/filter.go, right?:
```golang
func (s *precomputedSum[N]) Aggregate(value N, attr attribute.Set) {
	s.set(value, attr)
}
```

`set` has this implementation ([here](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/metric/internal/sum.go#LL35-L39C2): 

```golang
func (s *valueMap[N]) set(value N, attr attribute.Set) { // nolint: unused  // This is indeed used.
	s.Lock()
	s.values[attr] = value
	s.Unlock()
}
```

So I think it was resetting the precomputed one each time Aggregate was called?

Reply:
> So I think it was resetting the precomputed one each time Aggregate was called?

Correct, which is the specified behavior when the attribute set is not filtered, if a user reports a value multiple times for the same attribute set (which they shouldn't, but if they do) the last one needs to be used. When an attribute filter is used and it reduces a set to a previously recorded set, [that value needs to be added to the previous one](https://github.com/open-telemetry/opentelemetry-specification/issues/2905). Which is what is being done in these changes. Or, that is the intention. Do you not see that happening?

Reply:
My reading of the old code was that it _always_ replaced the current value.  My reading of the new code is that it _always_ adds the new value to the current value.  If I have your test case:
```golang
tr, err := mtr.AsyncFloat64().Counter("afcounter")
if err != nil {
	return err
}
_, err = mtr.RegisterCallback([]instrument.Asynchronous{ctr}, func(ctx context.Context) {
	ctr.Observe(ctx, 1.0, attribute.String("foo", "bar"), attribute.Int("version", 1))
	ctr.Observe(ctx, 2.0, attribute.String("foo", "bar"))
	ctr.Observe(ctx, 1.0, attribute.String("foo", "bar"), attribute.Int("version", 2))
})
```

And your view:
```golang
WithView(NewView(
	Instrument{Name: "*"},
	Stream{AttributeFilter: func(kv attribute.KeyValue) bool {
		return kv.Key == attribute.Key("foo")
	}},
)),
```

The first time Collect is called, I would (correctly) get a metric with value 4 (1 + 2 + 1).

But the second time Collect is called, wouldn't I (incorrectly) get a metric with value 8 (where the correct value is still 4)?

Reply:
Gotcha, thanks for the breakdown. I think I see your point now.

I'm AFK currently, but will look into this tomorrow.

Reply:
Updated, please take another look.

I think I have resolved this concern, and added tests to validate the solution. I'm going to resolve this, but if you think I missed something please unresolved.